Docs on the GUID-GUID recommender

dzeber opened the pull request:
Adding some doc on the logic underlying the GUID-GUID recommender, treatments, graph representation etc.

I've created two doc pages:

- `GuidGuidRecommender.md`: description of the `GuidGuidCoinstallRecommender` at a fairly high level, making the connection to implementation
- `GraphRepresentation.md`: more technical details about the graph representation of the coinstallation data and recommendation algorithm. I've added detailed descriptions of each of our implemented treatments here as well. This is more for our own reference as we start thinking about graph-based evaluation metrics.

There is also some minor cleanup of the main README.

r? @mlopatka @birdsarah

Commit:
Fix typos and minor cleanup in README

Commit:
Start adding doc on the GUID-GUID recommender

Commit:
Minor reorg of the recommender doc

Commit:
Reworked the current recommender doc

Commit:
Some more docs reorg:

- Moved technical details to the end
- Added graph-based interpretation of full recommendation algorithm

Commit:
Started adding doc on normalization treatments

Commit:
Added doc on RowNormSum normalization

Commit:
Added doc on graph pruning treatments

Commit:
Added doc on selecting recommendations as a treatment

Commit:
Moved graph representation details to seperate file

Commit:
Added description of graph properties

mlopatka commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+as much of the intuition behind the treatments
+is drawn from this representation.
+
+An add-on relational graph is a directed graph $G = (V,E)$ in which:
```
the pure co-installation graph can be represented (and thought of) as a purely undirected graph.

mlopatka replied:
I do see the note further down, but we should be as clear as possible at this moment wrt to the graph properties.

mlopatka commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+- A relational graph is induced directly by the coinstallation data,
+    considering add-ons related if they are coinstalled,
+    and using coinstallation counts as weights.
```
co-installation counts, or some statistic derived from the co-installation counts as weights.

mlopatka commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+### Recommendation graph
+
+The recommendation graph is obtained by dropping all edges except those
+leading to each add-on's top N most relevant recommendations,
```
as such it is necessarily a strict subset of __treated_graph__

mlopatka commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+Although these normalizations are described
+in terms of the raw coinstallation matrix,
+they can be applied as treatments to any add-on relational graph.
```
and unless otherwise stated, may be sequentially applied as part of a graph transformation workflow with an arbitrary number of treatment steps.

dzeber replied:
I've added a comment about this in the higher-level Treatments section.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+and relevance scores are given by the edge weights.
+In this setting:
+
+- A relational graph is induced directly by the coinstallation data,
```
Is induced a technical term with a specific meaning?
* If yes, good to know.
* If no, it feels jargon-y / technical. "created / built from" would suffice.

dzeber replied:
Good catch. It is a technical term, but it means something different from the way I've used it here.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+- A relational graph is induced directly by the coinstallation data,
+    considering add-ons related if they are coinstalled,
+    and using coinstallation counts as weights.
+- Treatments can be thought of as transformations
```
I would start with "We define the concept a Treatment. A Treatment object has a `treat` method that accepts a graph and return a modified graph." You could then continue about there being different general classes of treatments.

dzeber replied:
"Treatment" here is supposed to refer back to the concept described in the GUIDGUIDRecommender docs, and I've linked this in. I plan to add API details in the Treatment section further down.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+- Treatments can be thought of as transformations
+    which modify add-on relational graphs by
+    adding or deleting edges, or adjusting the edge weights.
+- The final selection of recommendations can itself be viewed as a treatment,
```
Feels a little extraneous.

dzeber replied:
Since we studied recommendations (looking for cycles) in terms of their graph representation, I thought this tied everything together nicely. Getting from the treated graph to the recommendation graph is also a graph transformation, and I felt like keeping that transformation isolated from the other "treatment" transformations made things less clear. I'm open to suggestions, though.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+    an edge from A to each of A's N recommendations.
+
+The recommendation algorithm can be rephrased
+in terms of this representation as follows:
```
"this" isn't clear in this sentence.

dzeber replied:
It was meant to refer to the bullet points just above. I thought using "this" would avoid repeating "graph representation".

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+2. Apply each of the specified treatments in turn to the coinstallation graph,
+    resulting in the __treated graph__.
+3. Apply the recommendation selection treatment to the treated graph
+    to produce the __recommendation graph__.
```
This 1, 2, 3 is my favorite bit so far.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+and entry $C_{ij}$ contains the weight for edge $(i,j)$
+if these two add-ons are connected in the graph, or 0 otherwise.
+The initial coinstallation dataset is in fact stored
+as a sparse representation of its adjacency matrix.
```
It is? I'm feeling out of my league in this paragraph with respect to graph theory. Can you clarify for me why the coinstallation json blob is "a sparse representation of its adjacency matrix"? (Or feel free to point me to a reference).

dzeber replied:
The "sparse" part just means that we interpret the JSON data as `{<row_id>: {<column_id>: weight}}`, listing all non-zero entries in the matrix (ie. all pairs of add-ons that will be connected by an edge in the graph). The [adjacency matrix](https://en.wikipedia.org/wiki/Adjacency_matrix) is a standard graph representation which is really useful in that it opens up the entire linear algebra toolkit to graph theory, eg. eigen-decompostion, etc.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+Since this graph is undirected, the adjacency matrix is symmetric
+($C_{ij} = C_{ji}$ for any pair $i,j$).
+Also, the row sum (or column sum, by symmetry) for add-on $i$ gives
+its overall number of installs across all profiles considered in the dataset.
```
What is a profile?

dzeber replied:
You've astutely hit on one of the big questions in Firefox data collection :). A profile is a "user" for our purposes, ie. each client_id corresponds to a profile. These are the same profiles you can switch between using the Firefox profile manager. All of our Telemetry is collected on a per-profile basis, and this is our closest approximation to a user. For a thrilling journey into this world, take a look at [this page](https://docs.telemetry.mozilla.org/concepts/profile/realworldusage.html).

I've used this term a few times in the GUID-GUID doc as well. I'll change each instance to "Firefox profile" for clarity.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+    the overall number of add-ons in the graph.
+    In other words, every vertex has at least one incident edge.
+- Edge weights correspond to the raw coinstallation count.
+- It may or may not be connected.
```
A ref to the definition of connected (e.g. https://en.wikipedia.org/wiki/Connectivity_(graph_theory)#Connected_graph) would be nice for noobs like me, but I can also look it up.

dzeber replied:
I added a link to a glossary at the beginning of the section.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+The result of applying the treatments to the coinstallation graph
+has these properties:
+
+- It is directed, as treatments generally do not operate on edges symmetrically.
```
This is too strong. I can think of plenty of treatments that would leave it undirected e.g. pruning, random selection.

dzeber replied:
I softened this a bit. Some treatments could leave it undirected, but even pruning could operate asymmetrically by removing an edge in one direction but not another (eg. dropping low-scoring recommendation candidates).

The idea is that, in general, the treated graph is directed. In some cases it could be undirected, but that's still a special case of directed, so any properties we discuss would still apply.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+- Both vertex in-degree and out-degree may take values from 0 to $|V|$,
+    and may be different.
+    In general, a treated vertex may have any number of incident edges.
+- Edge weights represent general relevance scores, which may not be symmetric.
```
This clear, and important, if there's a place to repeat and elevate it, great.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+- Edge weights represent general relevance scores, which may not be symmetric.
+    The weight on edge (A,B) represents the relevance of B in relation to A,
+    which may be different from the relevance of A in relation to B.
+- It may or may not be strongly or weakly connected.
```
Definition here: https://en.wikipedia.org/wiki/Connectivity_(graph_theory)#Definitions_of_components,_cuts_and_connectivity

* weakly connected  "A directed graph is called weakly connected if replacing all of its directed edges with undirected edges produces a connected (undirected) graph.
* connected "It is connected if it contains a directed path from u to v or a directed path from v to u for every pair of vertices u, v."
* strongly connected  "It is strongly connected, diconnected, or simply strong if it contains a directed path from u to v and a directed path from v to u for every pair of vertices u, v. The strong components are the maximal strongly connected subgraphs. "

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+    A is a recommendation for B.
+- Vertex in-degree may take values from 0 to $|V|$,
+    meaning that any number of add-ons may recommend the current vertex add-on.
+    Vertex out-degree is constrained to be at most N, representing the add-ons
```
I think this should be one bullet out.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+    meaning that any number of add-ons may recommend the current vertex add-on.
+    Vertex out-degree is constrained to be at most N, representing the add-ons
+    recommended for the current vertex add-on.
+- Edge weights no longer play a role.
```
I though edge weights are the order of the recommendations. I was assuming we would still be returning an ordered list.

I can also see the "this is the bucket of N recommendations" point, I just hadn't been thinking about it that way.

dzeber replied:
I'm thinking of the recommendation graph we worked with when we were looking for cycles. It was unweighted, because of the "bucket of recommendations" intuition.

This is related to our discussion in #42 where we decided to drop the scores. The list returned by the `recommend` function is ordered, but in the graph representation, there's no concept of ordered but unweighted edges.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+the edge weights may be modified asymmetrically,
+meaning that a single undirected edge in the original graph
+may get transformed into two directed edges with different weights
+in the result.
```
Very helpful.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+This treatment accounts for the popularity of an add-on in terms of
+how widely it is coinstalled,
+ie. how many different add-ons it appears coinstalled with,
+under the assumption that the more widely coinstalled an add-on is,
```
This sentence is a bit long and might be clearer as follows:

This treatment accounts for the popularity of an add-on in terms of how widely it is co-installed. 

It re-weights / treats the graph based on the hypothesis: the more widely co-installed an add-on is the less likely it is to be a relevant recommendation.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+in the result.
+
+Although these normalizations are described
+in terms of the raw coinstallation matrix,
```
My spell-checker wants coinstallation, coinstall etc to be hyphenated. I'm not sure. You have been consistent.

dzeber replied:
I've been using "coinstallation" mainly as a stylistic choice to avoid having hyphens all over the place. (On the other hand, the official Mozilla term "add-on" is hyphenated, so I've also used that consistently :)).

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+with similar total installation counts,
+of which some may be coinstalled more widely than others.
+
+This treatment is implemented as [`RowCount`](../taar_lite/recommenders/treatments.py#L87).
```
I think a simple diagram for each of the normalizations might be illustrative.

Here's one I made for RowCount

```
coinstalls = {
    "A": {"B": 100, "C": 100, "D": 100, ......},
    "B": {"A": 100, "C": 5, "D": 1},
    "C": {"A": 100, "B": 5},
    "D": {"A": 100, "B": 1},
    ....... continuing where A is coinstalled with many, many addons
}

counts = {
    'A': 100,  # A is co-installed with 100 add-ons
    'B': 3,  # B is co-installed with 3 add-ons
    'C': 2,  # C is co-installed with 2 add-ons
    'D': 2,  # D is co-installed with 2 add-ons
}

# Every number in the co-install matrix is then divided by the number in the counts matrix.

after_treatment = {
     'A': {'B': 33.33, 'C': 50.0, 'D': 50.0, .....},
     'B': {'A': 1, 'C': 2.5, 'D': 0.5},
     'C': {'A': 1, 'B': 1.67},
     'D': {'A': 1, 'B': 0.33},
    ......
}
```
You can see how A has been down-weighted due to it's global popularity and the interplay between B, C, and D rises to the surface.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+This treatment is implemented as [`RowCount`](../taar_lite/recommenders/treatments.py#L87).
+
+
+### Total relevance normalization
```
I'm concerned with overloading the word relevance.

Aren't all of the treatments various views on the relevance? Or does relevance mean something specific that I don't know (likely).

dzeber replied:
I'm using "relevance of B to A" to mean the weight on the edge from A to B. This normalization divides by the total relevance of B to all other add-ons, ie. the sum of weights on all edges entering B.

I've tried to make this more clear by adding a definition in the top section, along with more details about the asymmetric case.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+This treatment accounts for the popularity of an add-on in terms of
+its overall total number of installations,
+under the assumption that the more widely used an add-on is,
+the less likely it is to be relevant.
```
As before, this sentence can be simplified.

In addition, it is only one word different from the previous hypothesis and this should probably be emphasized. Maybe bolding *used* vs *co-installed*.

Also the distinction that you mean co-installed as a Y/N count is subtle and not necessarily obvious if you were coming to this for the first time.

I think elsewhere we say co-installed to mean the number of co-installs.

"used" is also somewhat problematic as a word, because we have no usage information, only co-installations counts.

(which reminds me - does telemetry let us know whether addons are enabled or not? - opened #50)

dzeber replied:
Agreed, this was unclear and confusing. I've adjusted the wording to hopefully make it more clear, and I've bolded the assumption statement in each treatment description.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+the proportion of B's total installs contributed by profiles
+that also have A installed.
+Thus, the highest-scoring add-ons B are those which are more likely
+to be coinstalled with A than with other add-ons.
```
I'm not quite getting this sentence.

Do you mean: Thus, the highest scoring add-ons *for* B are those which are more likely to be co-installed with A more than .... no that doesn't make sense... Can you update this.

dzeber replied:
I meant "the highest scoring add-ons B that are related to A". I've updated it.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+For a general add-on relational graph, the matrix may no longer be symmetric,
+but the intuition behind the normalization still applies.
+In this case, we can view the column sum for B
+as an __aggregate relevance score__ over the universe of add-ons,
```
okay, definitely feeling like i don't know the meaning of relevance. can you point to a definition for fellow noobs like me.

dzeber replied:
As mentioned in a previous comment, I've added a definition of relevance in the top section.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+in relation to each other add-on A.
+The normalization converts raw relevance scores to
+the proportion of aggregate relevance derived from add-on A.
+
```
Again, I'd like a diagram.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+### Proportional total normalization
+
+This treatment is a rescaled version of the [total relevance normalization](#total-relevance-normalization).
```
rescaled -> re-scaled 

re-scaled is the same as re-weighted right? (just checking my understanding)

dzeber replied:
Yes, it means divided by something different. AFAIK "rescale" is in the dictionary :).

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+Given add-on A, the coinstallation count for (A,B) for each add-on B
+is first divided by the sum of all A's coinstallation counts.
+The total relevance normalization is then applied to the resulting proportions.
```
So this is RowCount / RowSum?

If yes, I think there's an extra case for tightening up language, and making an effort to be repetitive so that these patterns pop-out for the reader.

If no, I'm sorry I'm not understanding this well and I'm happy to chat.

dzeber replied:
> So this is RowCount / RowSum?

Not exactly, since RowCount uses the number of non-zero entries to normalize. This one divides by row sum and then divides by column sum, so it's more like applying RowSum along the rows and then along the columns.

We should really rename the treatment implementations, as I think the Row* naming scheme is one of the main sources of confusion (and why I gave them more descriptive names in the doc).

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+relevance budget to 1.
+The next step then normalizes each coinstalled add-on's
+resulting aggregate relevance to 1.
+
```
Again, a worked out example will probably help folks like me.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+Another class of treatments we may wish to apply involves modifying
+the vertices or edges of the relational graph themselves,
+possibly as a function of the edge weights.
```
The other treatments already mentioned are also modifying the edges of the graph as a function of edge weights.

What about changing modifying -> removing (aka pruning)

dzeber replied:
I was thinking along the lines of the other treatments leaving the vertices and edges alone and only modifying the weights. I changed this to "sets of vertices or edges"

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+in that they only depend on the current add-on,
+we are interesting in optimizing the quality of experience
+for such users as well,
+such as not returning to a previous add-on in a short number of steps.
```
Everything from L271 (Also we may wish...) to end of paragraph is super important and useful context but is not specific to pruning. Move it up to initial pre-amble.

dzeber replied:
Hmm, I don't feel this really belongs anywhere else, since this is mainly accomplished by "pruning". I've left it in place and added a sentence to highlight this connection, but I'm happy to discuss further.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+such as not returning to a previous add-on in a short number of steps.
+
+
+### Minimum relevance pruning
```
In general, I think it would be helpful to put the links to the implementations at the top of sections. This will help reader know whether a paragraph is pre-amble or a discussion of a specific existing implementation.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+### Minimum relevance pruning
+
+This treatment removes add-on vertices whose relevance score is lower
```
Can we replace `relevance` with `total co-installations`?

dzeber replied:
As discussed above, I've tried to use relevance consistently to mean the edge weights. The idea was that sometimes these might not be the coinstallation counts, eg. when we chain treatments together.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+than a given threshold,
+under the assumption that candidates with such low relevance
+would not make good recommendations, even if there no candidates
+with higher relevance.
```
My understanding for the rationale was that it's privacy preserving.

dzeber replied:
I added a comment about this. I'm reluctant to highlight the privacy aspect too much since it's not really a quantifiable guarantee.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+The relevance threshold is computed based on an auxiliary list
+of overall relevance scores for each add-on.
+It is chosen as 5% of the mean of all the overall scores.
```
I think "5% of the mean of the total coinstallation for each addon" or something similar might be clearer. "Overall scores" feels like introducing more language.

dzeber replied:
This refers to the "relevance scores". I think this should be pretty clear, since this terminology is used a few times in the paragraph.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+This treatment is implemented as [`MinInstallPrune`](../taar_lite/recommenders/treatments.py#L36).
+The auxiliary score list is supplied as the keyword arg `ranking_dict`,
+a dict mapping add-on GUIDs to overall relevance scores.
+
```
Again a diagram.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+### Recommendation selection
+
+The procedure for [selecting recommendations](./GuidGuidRecommender.md#selecting-recommendations)
+it itself just another graph pruning treatment.
```
it -> is

I would have said a "sorting and pruning" treatment.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+
+For each add-on A in the graph, the edges leaving A are ordered
+by decreasing relevance score (weight),
+with ties broken using a supplied list of auxiliary add-on scores.
```
Ties could be broken in a bunch of ways. Is this a general paragraph or an implementation description?

dzeber replied:
It's an implementation description, as are all the treatment descriptions in this doc. The general overview of the recommendation process is in the GuidGuidRecommender doc.

birdsarah commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+we ignore the final edge weights, and can optionally set them to all to 1.
+
+
+# Quality and health metrics
```
Don't put in placeholder headings for content that doesn't exist yet. 

Just add the content when we have the content.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+The recommender system is implemented by the [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10) class,
+which is instantiated by the [TaarLiteAppResource](../taar_lite/app/production.py#L44).
+
+Further technical details are provided on [this page](./GraphRepresentation.md).
```
perhaps "further details of the indivudial steps are provided on [this page]"

birdsarah replied:
Also, would move this up to L25, straight after the steps.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+The GUID-GUID coinstallation recommender system generates recommendations
+for add-ons which are similar to a given add-on,
+where similarity is based on the number of profiles
+which have both add-ons installed.
```
We're into taste territory. To my eyes this is a little grandiose.

The GuidGuidCoinstalationRecommender Class generates recommendations, based on the co-installation rates of add-ons.

dzeber replied:
I'd prefer to keep this as just a general introduction to what this is about. I feel like it's clear, if a bit long.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+As recommendations are only surfaced
+for a specific set of add-ons meeting certain criteria,
+and depend only on characteristics of the add-ons themselves,
+recommendation sets for each add-on can be (and are) computed in advance.
```
Bit long. What about: 

Recommendations are only surfaced based on the input-data, rather than run-time or user data, and therefore can be (and are) computed in advance.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+and the quality of recommendations.
+
+The recommender system is implemented by the [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10) class,
+which is instantiated by the [TaarLiteAppResource](../taar_lite/app/production.py#L44).
```
I don't think this sentence is necessary, and it's a little confusing given that app/production instantiates multiple instances of `GuidGuidCoinstallRecommender`. I really don't think we need to mix in deployment concerns to these docs.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+# Input data
+
+The primary data source used in generating recommendations
+is the file `guid_coinstallation.json`,
```
hmmm....`GuidGuidCoinstallRecommender` doesn't care what you call the file.

I'd rather separate out the general from the deployment specifics.

So, I'm interested in the format of the input data that GuidGuidCoinstallRecommender expects and what it means.

That, in production it comes from an ETL job, and is lazily loaded through a cache every 4 hours, etc, etc belongs in full documentation about the deployment / production instance.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+Coinstallation counts are computed based on the Telemetry data
+submitted by a random sample of recently active Firefox profiles.
+For each pair of add-ons in the list,
+we count the number of profiles in which both add-ons are installed and enabled.
```
Cut this.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+mapping coinstalled GUIDS to their coinstallation count,
+dropping zero counts.
+On initializing a [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10),
+it is supplied as `raw_coinstall_dict`.
```
Keep this, but add a little code-snippet showing the dictionary format to make it super clear.

Why is "sparse matrix representation" in quotes? (And for my own learning, I'm going to ask you later about why this is a sparse matrix representation. I agree it's a sparse matrix, but I thought the representation, the actual data we got, was dense....maybe it's not, maybe it's full of 0s...if that's true, let's update code examples including tests to reflect that).

dzeber replied:
I guess the quotes aren't necessary. It's just one of those things where there are many things that can be called "sparse matrix representation", so it's not really a formal definition.

The full matrix that has a row and column for each GUID is sparse with lots of 0s. The representation is "dense" in the sense that we're dropping all the 0s for efficiency, but the underlying matrix it's representing is sparse. :)

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+it is supplied as `raw_coinstall_dict`.
+
+Using this dataset as the basis for recommendations
+entails the following assumptions:
```
So picky, but the bullet points are not all assumptions - but they are all important.

dzeber replied:
True, I missed the mark on word choice. I think "constraint" is more appropriate.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+    (otherwise they would not appear in the file at all).
+- Profiles with only a single add-on installed do not contribute to the counts.
+
+__TODO:__ How are system add-ons handled? Add a note about this.
```
Open an issue for this TODO

dzeber replied:
Ah, they are specifically excluded: https://github.com/mozilla/python_mozetl/blob/4bd8ceed909ae779826c383cbf7659b5dbf5869b/mozetl/taar/taar_lite_guidguid.py#L63

In fact, I don't think system add-ons would be listed on AMO either.

dzeber replied:
I'll add a comment about this in the README section on that script instead.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+On initializing a [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10),
+it is supplied as `tie_breaker_dict`.
+It is also used as the `ranking_dict` passed to the [`MinInstallPrune`](../taar_lite/recommenders/treatments.py#L36) treatment.
+
```
Again, separate out to the general Object related concerns only. 

Also, it's probably worth a note explaining why the ranking dict cannot be computed from the co-install dict. Just to be extra clear. (Answer being ranking dict lets us get the popularity of addons that are only installed by one client, and that's it I think, maybe there's more)

birdsarah replied:
It's perhaps worth noting that I deliberately called the recommender attribute `tie_breaker_dict` because anything could be used to break ties. 

The fact that the same data is used as an input to `MinInstallPrune` is coincidental and I wanted to capture that abstraction.

birdsarah replied:
When tie_breaking aka ordering becomes just another treatment, that will be extra clear, and there will be no `tie_breaker_dict` input to the recommender, it'll just be an input to another treatment, so I would suggest removing these last two paragraphs as they'll go soon anyway and we'll add a "Deterministic Ordering" section to the Treatments page.

dzeber replied:
I've updated these paragraphs to address your points. I think I'll leave them in for now, as they're already there, and remove them as a part of the PR that converts tie breaking to a treatment.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+
+The TAAR-lite recommender is built on the following underlying principle:
+_add-ons which tend to be installed together are similar or complimentary,
+and the number of coinstallations indicates the strength of this relationship._
```
I would say they're much more likely to be complimentary than similar.

I somehow wanted this to be broader e.g. about the fact that it's super anonymous data, but can provide valuable add-on recommendations.....but I can't come up with anything succinct and I don't think i'ts' necessary.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+- pruning out obscure add-ons
+- better balancing relevance against diversity
+- improving the experience for users clicking through a chain of recommendations
+    on AMO, such as pruning out cyclical recommendations.
```
I really liked this.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+referred to as __treatments__,
+to the raw coinstallation data.
+As recommendations are based solely on the treated data,
+we are assuming that it encapsulates all the information
```
it = what?

dzeber replied:
it = "the treated data". Maybe it'll be clearer if I change "it" to "this".

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+[here](./GraphRepresentation.md#treatments).
+
+Treatments are implemented as extensions of the [`BaseTreatment`](../taar_lite/recommenders/treatments.py#L11) class,
+which apply a transformation via the `treat()` method.
```
I don't think this is necessary.

The change of gear from super high-level to low-level was jarring.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+Treatments are implemented as extensions of the [`BaseTreatment`](../taar_lite/recommenders/treatments.py#L11) class,
+which apply a transformation via the `treat()` method.
+On initializing a [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10),
+a list of `BaseTreatment` instances are supplied as `treatments`.
```
But I would keep this.

birdsarah commented on docs/GuidGuidRecommender.md:
```diff
@@ -0,0 +1,144 @@
+
+Recommendations are requested from a `GuidGuidCoinstallRecommender` instance
+using the [`recommend()`](../taar_lite/recommenders/guidguid.py#L123) method.
+The list of tie-breaking scores was supplied on initialization as `tie_breaker_dict`.
```
I would suggest cutting from here to the end - see previous comments about `tie_breaker_dict` and deployment vs class concerns.

birdsarah reviewed (changes requested):
This is GREAT!

I know I made a lot of comments, but I wouldn't have been able to without all this work.

My comments area a mix of:
* styling / sentence things which are totally optional
* questions I have
* and things I think I'd like to see changed

I think overall the big picture things I'd like to see changed are:

1. obsessively consistent use of terms, and ideally a definitions section. This is to help people like me who are new and really don't know whether words are interchangeable and so it reduces the cognitive load. 
2. Mini examples for each of the treatments for people like me who are more concrete / visual thinkers
3. Do not mix-in production / deployment details. They can be written up in a seperate doc about app/production.

mlopatka reviewed (approved):
Super nice PR! We really needed to get good docs around this. I've made a few comments, mostly nits.

crankycoder commented on docs/GraphRepresentation.md:
```diff
@@ -0,0 +1,321 @@
+However, since the relation of being coinstalled is symmetric,
+the initial coinstallation graph is undirected.
+
+A relational graph has an associated __adjacency matrix__ $C$,
```
Dumb question - is markdown supposed to render the $ and curly brace syntax in a special way?

dzeber replied:
This is math notation in Latex. It is supposed to get rendered as a nicely formatted math expression, which is usually done using [MathJax](https://www.mathjax.org/).

It's not a standard part of Markdown, and needs to be supported by the Markdown flavour. I don't think it's supported by Github's automatic rendering, but I used it anyway because it's a good way to write precise expressions in text, and it's pretty readable as text. I could instead surround it with backticks or something so that it stands out.

Commit:
Updated GuidGuidRecommender.md in response to review

Commit:
Further updates to README:

- added details on filtering done in `taar_lite_guidguid.py`
- removed point on `taar_lite_guidguid.py` removing rare combinations

dzeber commented:
@birdsarah Thanks for taking the time to review this in depth - I appreciate the detailed comments, and I think they've really helped me hone down the focus.

I've made the requested changes to `GuidGiudRecommender.md`--updates to the other doc will follow.

I've also added a few details to the README about the `taar_lite_guidguid.py` processing. One thing I wanted to call out is that I didn't see any pruning of rare combinations in there (I think we'll be doing this as a recommender treatment instead), so I removed that bullet point.

Commit:
Updates to GraphRecommender.md in response to review (part 1)

As well as responding to the specific review comments, several uses of
terms defined in the GuidGuidRecommender doc have been linked to the relevant
sections.

Commit:
Updates to GraphRecommender.md based on review (part 2)

Commit:
Updates to GraphRecommender.md based on review (part 3)

Addressed all comments except for adding diagrams.

Commit:
Changed 'total installs' to 'total coinstalls' in description of total relevance normalization.

This reflects the fact that total coinstallations may be different
from total installations.

Commit:
Added example for add-on count normalization

Commit:
Added example for total relevance normalization

Commit:
Started to add an example for the proportional total normalization

Commit:
Finished example for proportional total normalization

- Changed to a different, more illustrative dataset
- Added a paragraph in the description describing how the row norm
  corrects for scale differences
- Added total relevance norm result in example for comparison

Commit:
Added example for min relevance pruning treatment

Commit:
Added example for recommendation treatment

Commit:
Added link to code for running examples

dzeber commented:
@birdsarah @mlopatka @crankycoder Thanks everyone for the detailed review! I believe I've addressed every comment, either through changes or by replying.

In response to Sarah's overall requests:

> 1. obsessively consistent use of terms, and ideally a definitions section. This is to help people like me who are new and really don't know whether words are interchangeable and so it reduces the cognitive load.

AFAICT I've used terminology consistently throughout, and relevant terms are now all defined (and usually bolded). Definitions are often inline in the text as opposed to in a formal glossary.

> 2. Mini examples for each of the treatments for people like me who are more concrete / visual thinkers

I've added an example for each treatment, some with graph diagrams. ASCII art FTW! I'd love to find a better way to create diagrams, but ideally it would render nicely in Github markdown without relying on external tools.

> 3. Do not mix-in production / deployment details. They can be written up in a seperate doc about app/production.

Done - these have been removed and parked for future use.
